### PR TITLE
Decentralize plugin security

### PR DESCRIPTION
**Pull request by rmuir**
Today ES core defines all the rules for plugins and special exceptions by name. But this means if a plugin developer needs a similar thing (perhaps just a temporary hack or workaround: maybe in a third party dependency of theirs), it is not possible today, and the only solution is to disable securitymanager entirely.

This is bad for a few reasons:
1. causes core developers stress because they are in the loop for all problems
2. encourages disabling security entirely.

This PR presents an alternative, more like the android model, where plugins can include an optional `plugin-security.policy` file, and users have to confirm the additional requested permissions on installation (if stdin is a controlling terminal, and there is also a new explicit `-b` batch flag for `bin/plugin` too just for completeness):

```
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@     WARNING: plugin requires additional permissions     @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
* java.lang.RuntimePermission accessClassInPackage.sun.reflect
* java.lang.RuntimePermission closeClassLoader
* java.lang.RuntimePermission createClassLoader
* groovy.security.GroovyCodeSourcePermission /untrusted
See http://docs.oracle.com/javase/8/docs/technotes/guides/security/permissions.html
for descriptions of what these permissions allow and the associated risks.

Continue with installation? [y/N]
```

I think its fairly safe, we protect plugins directories pretty well now (even unix filesystem permissions are improved recently). These permissions are _only_ granted to the plugin, like today, which means this stuff is contained pretty well. Of course it forces plugin authors to write proper security code, but I think that is ok, versus it not being possible at all. I added recommendations and an example to the plugin authors documentation to try to help with this. Even in the worst case if a plugin author adds `AllPermission` and screws up all the security code completely, its still better than having no security checking at all for ANY code... That is the major motivation for me wanting to do this, I think it will make things better. It just took me until now to figure out how to make the IDEs work :)

This change also allows tests to work from IDEs for such plugins without as many hacks as before, by configuring some special resources.

### CODE REVIEW

**Comment by dweiss** [comment]
Not only it looks like an awesome feature, but also a nice cleanup of existing code (I went through the diff out of curiosity). Very cool.

**Comment by clintongormley** [comment]
nice!

File: core/src/main/resources/org/elasticsearch/bootstrap/security.policy [hunk 1]
```diff
@@ -37,52 +37,6 @@ grant codeBase "${es.security.jar.lucene.core}" {
   permission java.lang.reflect.ReflectPermission "suppressAccessChecks";
 };
 
-//// Special plugin permissions:
```
Comment (s1monw) [inline core/src/main/resources/org/elasticsearch/bootstrap/security.policy]: boom! nice once - that 46 lines of deletions is worth the change IMO already

File: core/src/test/resources/org/elasticsearch/plugins/security/simple-plugin-security.policy [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+grant {
+  // needed to waste paper
```
Comment (s1monw) [inline core/src/test/resources/org/elasticsearch/plugins/security/simple-plugin-security.policy]: LOL

File: docs/plugins/authors.asciidoc [hunk 3]
```diff
@@ -119,3 +119,35 @@ You may also load your plugin within the test framework for integration tests.
 Read more in {ref}/integration-tests.html#changing-node-configuration[Changing Node Configuration].
 
 
+[float]
+=== Java Security permissions
```
Comment (s1monw) [inline docs/plugins/authors.asciidoc]: I wonder if we should add a discouraging paragraph here that makes folks to think twice about adding exceptions? I am not sure how  it should be phrased just putting the question out there...

**Comment by s1monw** [comment]
this looks awesome! thanks for doing this!

**Comment by dadoonet** [comment]
<3

File: docs/plugins/authors.asciidoc (hunk 3 above)
Comment (rmuir) [inline docs/plugins/authors.asciidoc]: I can reword it better.

**Comment by rmuir** [comment]
@s1monw I pushed changes to improve the docs. I also moved the check after 'plugin already exists' check, so that the user doesn't have to enter Y/N before that case.

File: core/src/test/resources/org/elasticsearch/plugins/security/simple-plugin-security.policy (hunk 2 above)
Comment (uboness) [inline core/src/test/resources/org/elasticsearch/plugins/security/simple-plugin-security.policy]: or for paper maps

**Comment by s1monw** [comment]
LGTM

**Comment by salyh** [comment]
Thats a great idea and i would love to see this as soon as possible released. I am working on a shield kerberos realm (nearly finished) and i had to disable security manager entirely cause all the <pre>Subject.doAs</pre> stuff is with current ES policy not allowed. Would also be nice to read krb5.conf directly from /etc/.

**Comment by jprante** [comment]
I hoped for plugin security, thanks! Right now I have to disable security manager completely to get my groovy-based web app plugin things going https://github.com/jprante/elasticsearch-webapp

**Comment by rmuir** [comment]
Hi @salyh @jprante ... I have set the feature for 2.2 because it depends on a ton of other work that was done in the 2.2 codebase. I'm afraid this is about the soonest we can get this out unfortunately.

Before releasing it I have in mind add some minor improvements to it (e.g. ability to use full policy syntax, possibly rename the configuration file to improve IDE/test support, etc). I am working on some of that at the moment.

**Comment by costin** [comment]
@rmuir Awesome stuff! Thanks for adding this in.

**Pull request closed without merging**